### ISSUE TITLE
Suggestions to get good results

### ISSUE THREAD

**Issue by juanmed** [hyc96/Real-time-photometric-stereo-and-surface-reconstruction-on-personal-devices#1]
Hi,

Thanks for sharing your code. 

I have been playing with photometric stereo for some time now and wanted to ask for your suggestions to get reasonable results (like shown in this repository). I have been unable to replicate those results. I tried with my face and with other objects, at various distances from the camera and screen, using a dark background and placing the camera either at the top and at the center of the screen, but with no success.

See for example the following images, each one for each lighting direction. They do not look that different:
![1](https://user-images.githubusercontent.com/21977733/78810611-869abd80-7a03-11ea-94fb-c23731e943c0.jpg)
![2](https://user-images.githubusercontent.com/21977733/78810673-9e724180-7a03-11ea-9867-49d70ae0a91d.jpg)
![3](https://user-images.githubusercontent.com/21977733/78810682-a0d49b80-7a03-11ea-9eb4-430c179c3c82.jpg)
![4](https://user-images.githubusercontent.com/21977733/78810691-a29e5f00-7a03-11ea-87e0-81e0fcd5c889.jpg)

And the result (using `python main.py 2`): 

![res_view1](https://user-images.githubusercontent.com/21977733/78810878-f315bc80-7a03-11ea-9693-d0e56ef7ed94.png)
![res_view2](https://user-images.githubusercontent.com/21977733/78810884-f4df8000-7a03-11ea-9229-29d01d4e78ca.png)

I was wondering if you could please comment about any step, setup or special care I am missing  to get reasonable results. 

Thank you!

**Comment by hyc96** [comment]
Did you get that rectangular shape from the soccer ball?

**Comment by juanmed** [comment]
> Did you get that rectangular shape from the soccer ball?

Yes, that is the result from the 4 images of the soccer ball. Each image corresponds to one light direction from the screen as obtained when executing `python main.py 1`.

**Comment by hyc96** [comment]
> > Did you get that rectangular shape from the soccer ball?
> 
> Yes, that is the result from the 4 images of the soccer ball. Each image corresponds to one light direction from the screen as obtained when executing `python main.py 1`.

That's bizarre. Can you show screencaps of other angles? Also the lighting condition seems bad if those are four images from different captures.

**Comment by juanmed** [comment]
> That's bizarre. Can you show screencaps of other angles? 

What do you mean with `screencaps of other angles`? To repeat the experiment with the ball in other position relative to the camera? relative to the screen? 

> Also the lighting condition seems bad if those are four images from different captures.

It does seem bad. The result was very similar with other objects (face, box):  the lighting result was not very different for each picture. I placed the object centered to the screen, at several distances from the camera/screen center ( from about 3cm up to about 15cms), with no reasonable results.

Could you please explain in detail what was the fixture you used for the result shown in the README file? (position of camera relative to screen, distance from face to screen, type of screen, type of background and distance from screen to background, any other relevant condition)

**Comment by hyc96** [comment]
Other angle of the reconstruction.

And it was just my colleague sitting in front of the laptop with lights off, nothing special.

**Comment by hyc96** [comment]
It could be the fact that the surface of your soccer ball violates Lambertian assumption too much

**Comment by juanmed** [comment]
Thanks for your comments. I will make the experiment again. In the meantime, Could you please share the input images you obtained for the result shown in the README file? I would like to get a feel of how different they are in terms of lighting, and reproduce the result shown. Thanks.

**Comment by juanmed** [comment]
Hi,

I repeated the experiments. I tried with a laptop with a brighter screen. For some tests, I got somewhat better results. See the following links:

- Travel Headrest: https://imgur.com/a/A0H5Gh5
- Hand: https://imgur.com/a/1ldC7bP
- Box: https://imgur.com/a/iGYBoqV
- Hand (Black Rubber Glove): https://imgur.com/a/cPe2NL2
- Ball 3: https://imgur.com/a/V9jIBTh
- Ball 2: https://imgur.com/a/kb20YUC

The best results were obtained with "Ball 3" and "Travel Headrest". It seems to me that the Lambertian reflectance assumption might be a reasonable for the ball. For the rest, the reconstruction is bad or, with some effort, something similar to the object  can be seen. The hand had the best result among the rest, at least the shape is recognizable, but the depth map is off. 

Moreover, it seems that in the `surface_normal` method, some misalignment is happening. All the reconstructed surfaces seem to have a straight line discontinuity that seems to be the edge of the image, and not a real line in the image. In the following crop it is clearer:

![line](https://user-images.githubusercontent.com/21977733/79067006-32d7ef00-7cf7-11ea-8f66-d68553bdf9eb.png)

That line seems to correspond to the boundary of the image. 

Based on these, do you have any ideas as to how to improve the results?

Thanks.

By the way, in the previous comment I was wondering if you could share the images from which you obtained the results in the README. Do you happen to still have them? I will appreciate it.

**Comment by hyc96** [comment]
Hi, 
Sorry I've been working on couple deadlines. I'll try to review these soon.
And I don't have them at the moment
